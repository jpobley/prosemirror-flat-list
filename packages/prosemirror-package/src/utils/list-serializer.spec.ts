import { describe, expect, it } from 'vitest'
import { setupTestingEditor } from '../../test/setup-editor'
import { ListDOMSerializer } from './list-serializer'

describe('ListDOMSerializer', () => {
  const {
    add,
    doc,
    p,
    bulletList,
    orderedList,
    uncheckedTaskList: taskList,
    expandedToggleList: toggleList,
    schema,
  } = setupTestingEditor()

  let editor: ReturnType<typeof add>

  it('can serialize list nodes into <ul>', () => {
    editor = add(doc(bulletList(p('A')), bulletList(p('B'))))

    const serializer = new ListDOMSerializer(
      ListDOMSerializer.nodesFromSchema(schema),
      ListDOMSerializer.marksFromSchema(schema),
    )

    const serialized = serializer.serializeFragment(editor.state.doc.content)

    expect(serialized.querySelectorAll('UL').length).toBe(1)
    expect(serialized.querySelectorAll('OL').length).toBe(0)
    expect(serialized.querySelectorAll('UL > LI').length).toBe(2)
    expect(serialized).toMatchSnapshot()
  })

  it('can serialize list nodes into <ol>', () => {
    editor = add(doc(orderedList(p('A')), orderedList(p('B'))))

    const serializer = new ListDOMSerializer(
      ListDOMSerializer.nodesFromSchema(schema),
      ListDOMSerializer.marksFromSchema(schema),
    )

    const serialized = serializer.serializeFragment(editor.state.doc.content)

    expect(serialized.querySelectorAll('UL').length).toBe(0)
    expect(serialized.querySelectorAll('OL').length).toBe(1)
    expect(serialized.querySelectorAll('OL > LI').length).toBe(2)
    expect(serialized).toMatchSnapshot()
  })

  it('can serialize list nodes with different types into a single <ul>', () => {
    editor = add(
      doc(
        bulletList(p('A')),
        taskList(p('B')),
        toggleList(p('C')),

        orderedList(p('D')),

        bulletList(p('D')),
        taskList(p('E')),
        toggleList(p('D')),
      ),
    )

    const serializer = new ListDOMSerializer(
      ListDOMSerializer.nodesFromSchema(schema),
      ListDOMSerializer.marksFromSchema(schema),
    )

    const serialized = serializer.serializeFragment(editor.state.doc.content)

    expect(serialized.querySelectorAll('UL').length).toBe(2)
    expect(serialized.querySelectorAll('OL').length).toBe(1)
    expect(serialized.querySelectorAll('UL > LI').length).toBe(6)
    expect(serialized.querySelectorAll('OL > LI').length).toBe(1)
    expect(serialized).toMatchSnapshot()
  })

  it('can serialize nested list node ', () => {
    editor = add(
      doc(
        bulletList(p('A'), orderedList(p('B')), orderedList(p('C'))),
        bulletList(p('D'), orderedList(p('E')), orderedList(p('F'))),
      ),
    )

    const serializer = new ListDOMSerializer(
      ListDOMSerializer.nodesFromSchema(schema),
      ListDOMSerializer.marksFromSchema(schema),
    )

    const serialized = serializer.serializeFragment(editor.state.doc.content)

    expect(serialized.querySelectorAll('UL').length).toBe(1)
    expect(serialized.querySelectorAll('OL').length).toBe(2)
    expect(serialized.querySelectorAll('UL > LI').length).toBe(2)
    expect(serialized.querySelectorAll('OL > LI').length).toBe(4)
    expect(serialized).toMatchSnapshot()
  })
})
