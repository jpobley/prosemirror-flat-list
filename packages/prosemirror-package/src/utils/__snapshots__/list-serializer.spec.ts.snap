// Vitest Snapshot v1

exports[`ListDOMSerializer > can serialize list nodes into <ol> 1`] = `
<DocumentFragment>
  <ol>
    <li
      class="prosemirror-flat-list"
      data-list-type="ordered"
    >
      <p>
        A
      </p>
    </li>
    <li
      class="prosemirror-flat-list"
      data-list-type="ordered"
    >
      <p>
        B
      </p>
    </li>
  </ol>
</DocumentFragment>
`;

exports[`ListDOMSerializer > can serialize list nodes into <ul> 1`] = `
<DocumentFragment>
  <ul>
    <li
      class="prosemirror-flat-list"
      data-list-type="bullet"
    >
      <p>
        A
      </p>
    </li>
    <li
      class="prosemirror-flat-list"
      data-list-type="bullet"
    >
      <p>
        B
      </p>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`ListDOMSerializer > can serialize list nodes with different types into a single <ul> 1`] = `
<DocumentFragment>
  <ul>
    <li
      class="prosemirror-flat-list"
      data-list-type="bullet"
    >
      <p>
        A
      </p>
    </li>
    <li
      class="prosemirror-flat-list"
      data-list-type="task"
    >
      <p>
        B
      </p>
    </li>
    <li
      class="prosemirror-flat-list"
      data-list-type="toggle"
    >
      <p>
        C
      </p>
    </li>
  </ul>
  <ol>
    <li
      class="prosemirror-flat-list"
      data-list-type="ordered"
    >
      <p>
        D
      </p>
    </li>
  </ol>
  <ul>
    <li
      class="prosemirror-flat-list"
      data-list-type="bullet"
    >
      <p>
        D
      </p>
    </li>
    <li
      class="prosemirror-flat-list"
      data-list-type="task"
    >
      <p>
        E
      </p>
    </li>
    <li
      class="prosemirror-flat-list"
      data-list-type="toggle"
    >
      <p>
        D
      </p>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`ListDOMSerializer > can serialize nested list node  1`] = `
<DocumentFragment>
  <ul>
    <li
      class="prosemirror-flat-list"
      data-list-collapsable=""
      data-list-type="bullet"
    >
      <p>
        A
      </p>
      <ol>
        <li
          class="prosemirror-flat-list"
          data-list-type="ordered"
        >
          <p>
            B
          </p>
        </li>
        <li
          class="prosemirror-flat-list"
          data-list-type="ordered"
        >
          <p>
            C
          </p>
        </li>
      </ol>
    </li>
    <li
      class="prosemirror-flat-list"
      data-list-collapsable=""
      data-list-type="bullet"
    >
      <p>
        D
      </p>
      <ol>
        <li
          class="prosemirror-flat-list"
          data-list-type="ordered"
        >
          <p>
            E
          </p>
        </li>
        <li
          class="prosemirror-flat-list"
          data-list-type="ordered"
        >
          <p>
            F
          </p>
        </li>
      </ol>
    </li>
  </ul>
</DocumentFragment>
`;
